pxd.c

```diff
@@ -1776,6 +1778,58 @@ static ssize_t pxd_inprogress_show(struct device *dev,
 	return sprintf(buf, "%d", atomic_read(&pxd_dev->ncount));
 }
 
+static int pxd_nodewipe_cleanup(struct pxd_context *ctx)
+{
+	struct list_head *cur;
+
+	if (ctx->fc.connected) {
+		return -EINVAL;
+	}
+
+	cancel_delayed_work_sync(&ctx->abort_work);
+	pxd_abort_context(&ctx->abort_work.work);
+
+	if (ctx->num_devices == 0) {
+		return 0;
+	}
+
+	spin_lock(&ctx->lock);
+	list_for_each(cur, &ctx->list) {
+		struct pxd_device *pxd_dev = container_of(cur, struct pxd_device, node);
+
+		disableFastPath(pxd_dev, true);
+	}
+	spin_unlock(&ctx->lock);
+
+	return 0;
+}
+
+static ssize_t pxd_release_store(struct device *dev,
+			struct device_attribute *attr, const char *buf, size_t count)
+{
+	static const char wipemagic[] = "P0RXR3l3@53";
```

Comment:
- Wouldn't it be more appropriate to extend the pxd_iotcl() to do the job?

More importantly can this be not done as a part of module unload? Same issue exists for stuck pxd IOs. The way nodewipe deals with them is by simply setting the timeout to 30secs. If you just add disableFastPath() to module cleanup, that should do it. Also this way px-storage UTs cleanup wouldn't need to do anything special.

Reply:
ioctl needs to open pxd_control and this is being interpreted as a userspace attach to px-fuse. So ioctl cannot be used from node wipe. If IOs are stuck to virtual pxd devices, they still do not hold direct references to backing devices, node wipe can proceed to pull the rug under them and will pass. But with fastpath active, there is a direct reference held that needs to be broken before node wipe can pass. They are not similar.

Reply:
No, ioctl has a reserved channel.
What do u think about the other option - doing it in module unload?

Reply:
Aborting IOs on abort timeout doesn't switch IO path... io path switch can happen only with  px-storage  being aware otherwise other targets will be unaware....  And while fastpath is active, IO being failed happens without any intervention..., this timeout run only ensures newer IOs do fail immediately.
Infact node wipe already uses timeoutDir to reset/force abort handler to run... extending it to release fastpath is more inline to current design.
Ioctl may work,  but need to aware of the proper device to open which may fail if module param pxd_num_contexts_exported gets used.

Can we not use the current posted method as is... i see no benefits vs other approaches to current approach.